### PR TITLE
JAMES-3382 Email query position

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 1]
```diff
@@ -1358,6 +1358,133 @@ trait EmailQueryMethodContract {
...
+  def resultsShouldStartAtThePositionProvidedByTheClient(server: GuiceJamesServer): Unit = {
+    val message: Message = Message.Builder
+      .of
+      .setSubject("test")
+      .setBody("testmail", StandardCharsets.UTF_8)
+      .build
+    server.getProbe(classOf[MailboxProbeImpl]).createMailbox(MailboxPath.inbox(BOB))
+    val otherMailboxPath = MailboxPath.forUser(BOB, "other")
+    val requestDate = Date.from(ZonedDateTime.now().minusDays(1).toInstant)
+    val messageId1: MessageId =  sendMessageToBobInbox(server, message, Date.from(requestDate.toInstant))
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1371]: extra space

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 2]
```diff
@@ -1358,6 +1358,133 @@ trait EmailQueryMethodContract {
...
+         |    },
+         |    "c1"]]
+         |}""".stripMargin
+
+    awaitAtMostTenSeconds.untilAsserted { () =>
+      val response = `given`
+        .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+        .body(request)
+        .when
+        .post
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1397]: Indent

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 3]
```diff
@@ -1358,6 +1358,133 @@ trait EmailQueryMethodContract {
...
+        .`then`
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1461]: Indent

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 4]
```diff
@@ -1484,6 +1484,67 @@ trait EmailQueryMethodContract {
...
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1525]: Indent

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
```
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryBackReferenceTest.loadingAccountWithBackReferences JSON documents are different:
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] Array "methodResponses" has different content. Missing values: [["Email/query",{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","queryState":"a6904f19","canCalculateChanges":false,"ids":["1","2","3"],"position":0,"limit":256},"c2"], ["Email/get",{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","state":"000001","list":[{"id":"3","subject":"message 3"},{"id":"2","subject":"message 2"},{"id":"1","subject":"message 1"}],"notFound":[]},"c3"]], extra values: [["error",{"type":"serverFail","description":"None (of class scala.None$)"},"c2"], ["error",{"type":"invalidResultReference","description":"Failed resolving back-reference: List((,List(JsonValidationError(List(MethodCallId(c2) references a MethodName(error) method),ArraySeq()))))"},"c3"]], expected: <[["Mailbox/query",{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","queryState":"33cc79f8","canCalculateChanges":false,"ids":["1"],"position":0,"limit":256},"c1"],["Email/query",{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","queryState":"a6904f19","canCalculateChanges":false,"ids":["1","2","3"],"position":0,"limit":256},"c2"],["Email/get",{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","state":"000001","list":[{"id":"3","subject":"message 3"},{"id":"2","subject":"message 2"},{"id":"1","subject":"message 1"}],"notFound":[]},"c3"]]> but was: <[["Mailbox/query",{"accountId":"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6","queryState":"33cc79f8","canCalculateChanges":false,"ids":["1"],"position":0,"limit":256},"c1"],["error",{"type":"serverFail","description":"None (of class scala.None$)"},"c2"],["error",{"type":"invalidResultReference","description":"Failed resolving back-reference: List((,List(JsonValidationError(List(MethodCallId(c2) references a MethodName(error) method),ArraySeq()))))"},"c3"]]>
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] 
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR] Errors: 
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsByCustomKeywordShouldReturnOnlyMailsWithThisCustomKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsBySystemKeywordShouldReturnOnlyMailsWithThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsBySystemKeywordShouldReturnOnlyMailsWithThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsBySystemKeywordShouldReturnOnlyMailsWithThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsBySystemKeywordShouldReturnOnlyMailsWithThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsBySystemKeywordShouldReturnOnlyMailsWithThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsInAFirstMailboxAndNotInTheSameMailboxShouldReturnEmptyResult » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsInAFirstMailboxAndNotSomeOtherMailboxShouldReturnMailsInFirstMailbox » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsNotByCustomKeywordShouldReturnOnlyMailsWithoutThisCustomKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsNotBySystemKeywordShouldReturnOnlyMailsWithoutThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsNotBySystemKeywordShouldReturnOnlyMailsWithoutThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsNotBySystemKeywordShouldReturnOnlyMailsWithoutThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsNotBySystemKeywordShouldReturnOnlyMailsWithoutThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsNotBySystemKeywordShouldReturnOnlyMailsWithoutThisSystemKeyword » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsReceivedAfterADateShouldBeExclusive » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsShouldBeSortedByDescendingOrderOfArrivalByDefault » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsShouldBeSortedInAscendingOrderByDefault » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsShouldBeSortedInAscendingOrder » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.listMailsShouldBeSortedInDescendingOrder » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldLimitResultByTheLimitProvidedByTheClient » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldListMailsInASpecificUserMailboxes » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldListMailsInAllUserMailboxes » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldListMailsNotInASpecificUserMailboxes » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldListMailsNotInZeroMailboxes » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldListMailsReceivedAfterADate » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldListMailsReceivedBeforeADateInclusively » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldListMailsReceivedBeforeADate » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldNotListMailsFromOtherUserMailboxes » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldReturnAnIllegalArgumentExceptionIfTheLimitIsNegative » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.shouldReturnIllegalArgumentErrorForAnUnknownSpecificUserMailboxes » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.theLimitshouldBeEnforcedByTheServerIfAGreaterLimitProvidedByTheClient » ConditionTimeout
[dacb1fd96640a2cdd9e763a888c8fc0d2a4996dc] [ERROR]   MemoryEmailQueryMethodTest.theLimitshouldBeEnforcedByTheServerIfNoLimitProvidedByTheClient » ConditionTimeout
```

Likely broke something :-)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 5]
```diff
@@ -1358,6 +1358,133 @@ trait EmailQueryMethodContract {
...
+    val request =
+      s"""{
+         |  "using": [
+         |    "urn:ietf:params:jmap:core",
+         |    "urn:ietf:params:jmap:mail"],
+         |  "methodCalls": [[
+         |    "Email/query",
+         |    {
+         |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |      "position": 1
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1387]: position: Int (default: 0) The zero-based index of the first id in the full list of results to return.

That is not clear in the test here (I fail to easily understand we have 2 mails in the DB)

**Review by Arsnael — COMMENTED** [review]
Read it

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 6]
```diff
@@ -1358,6 +1358,133 @@ trait EmailQueryMethodContract {
...
+           |                "limit": 256,
+           |                "canCalculateChanges": false,
+           |                "ids": []
+           |            },
+           |            "c1"
+           |        ]]
+           |}""".stripMargin)
+    }
+  }
+
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1486]: We miss a test with `0`

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala [hunk 7]
```diff
@@ -32,7 +32,7 @@ case class FilterCondition(inMailbox: Option[MailboxId],
                            hasKeyword: Option[Keyword],
                            notKeyword: Option[Keyword])
 
-case class EmailQueryRequest(accountId: AccountId, limit: Option[LimitUnparsed], filter: Option[FilterCondition], comparator: Option[Set[Comparator]])
+case class EmailQueryRequest(accountId: AccountId, position: Option[Position], limit: Option[LimitUnparsed], filter: Option[FilterCondition], comparator: Option[Set[Comparator]])
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala]: Did I miss the part about position parsing?

**Review by chibenwa — COMMENTED** [review]
I will do a quick follow up (and fixups) of this work.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala (hunk 5 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1387]: Took me a while to spot there were two messages as well... kept reading back and forth the specs ^^'

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Deployment tests

Can be considered :green_apple:

test this please anyway

Will need to be rebased once `limit` #3792 is merged

**Comment by chibenwa** [comment]
(just rebased)

**Comment by chibenwa** [comment]
test this please 

(blind - no VPN)

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**